test/unified/runner.js

```diff
@@ -0,0 +1,366 @@
+  }
+  static callbacksAll(tasks, cb) {
+    const results = [];
+    let foundError = undefined;
+    return tasks.reverse().reduce((cb, task) => {
+      return () => {
+        if (foundError) return cb(foundError);
+        task((err, result) => {
+          if (err) {
+            foundError = err;
+            return cb(err);
+          }
+          results.push(result);
+          return cb(undefined, results);
+        });
+      };
+    }, cb)();
+  }
+  /** @see https://gist.github.com/bisubus/2da8af7e801ffd813fab7ac221aa7afc#file-pick-es2015-js */
+  static pick(obj, picked) {
+    return Object.keys(obj)
+      .filter(key => picked.indexOf(key) >= 0)
+      .reduce((newObj, key) => Object.assign(newObj, { [key]: obj[key] }), {});
+  }
+  resetEntityMap(cb) {
+    this.entities = [];
+    this.clientEvents = [];
+    this.clientEntityIds = [];
+    return cb();
+  }
+  closeClients(cb) {
+    const cbs = this.clientEntityIds.map(id => {
+      return cb => {
+        const client = this.getEntity(id);
+        return client.close(true, cb);
+      };
+    });
+    return Runner.callbacksAll(cbs, cb);
+  }
+  static singleOrArray(value) {
+    if (!value) return [];
+    if (Array.isArray(value)) return value;
+    return [value];
+  }
+  static createMetadata(req) {
+    const topology = Runner.singleOrArray(req.topology);
+    const minServerVersion = req.minServerVersion;
+    const maxServerVersion = req.maxServerVersion;
+    let mongodb = [];
+    if (minServerVersion) mongodb.push(`>=${Version.clean(minServerVersion)}`);
+    if (maxServerVersion) mongodb.push(`<=${Version.clean(maxServerVersion)}`);
+    const requires = {};
+    if (topology.length) requires.topology = topology;
+    if (mongodb.length) requires.mongodb = mongodb.join(' ');
+    return { requires };
+  }
+  static createTestDescription(test, data) {
+    let description = [test.description];
+    if (data.topology && data.topology.length) description.push(data.topology.join(','));
+    if (data.mongodb) description.push(data.mongodb);
+    return description.join(' - ');
+  }
+  static handleSpec(spec) {
+    const r = new Runner(spec);
+    describe(r.spec.description, () => {
+      r.spec.tests.forEach(test => {
+        r.spec.runOnRequirements.forEach(req => {
+          const metadata = Runner.createMetadata(req);
+          const description = Runner.createTestDescription(test, metadata.requires);
+          it(description, metadata, function (done) {
+            Runner.callbacksAll(
+              [
+                cb => r.createTopClient(cb),
+                cb => r.createEntities(cb),
+                cb => r.removeInitialData(cb),
+                cb => r.createInitialData(cb),
+                cb => r.handleTest(test, cb),
+                cb => r.closeClients(cb),
+                cb => r.resetEntityMap(cb)
+              ],
+              done
+            );
+          });
+        });
+      });
+    });
+  }
+}
+
+module.exports = {
```

Comment:
One of the nice properties of the existing spec test runner is that we are able to model the spec-specific helper functions in a declarative way on a subclass of the `TestRunnerContext`. You can see examples of that here:
- [SDAM integration tests](https://github.com/mongodb/node-mongodb-native/blob/master/src/operations/command.ts#L71-L73)
- [Transactions](https://github.com/mongodb/node-mongodb-native/blob/master/test/functional/transactions.test.js#L19-L81)
- [Sessions](https://github.com/mongodb/node-mongodb-native/blob/3.6/test/functional/sessions.test.js#L161-L185)

Is there any plan to support similar modular support here?

Reply:
Here the examples are showing custom assertion methods, I had no need for these in the POC with the example spec provided, we can bring these in where / if they match with the specs assertion / matchers.